Add Descript-Audio-Codec model

kamilakesbi opened the pull request:
# What does this PR do?

This PR aims at adding [Descript-Audio-Codec model](https://github.com/descriptinc/descript-audio-codec/tree/main), a high fidelity general neural audio codec, to the Transformers library. 

This model is composed of 3 components: 

- An Encoder model. 
- A ResidualVectorQuantizer model, which is used with the encoder to obtain the audio quantized latent codes. 
- A Decoder model, used to reconstruct the audio after compression. 

This is still a draft PR. Here's what I've done for now: 

1. Adapted the model to Transformers format in `modeling_dac.py`. 
2. Added the checkpoint conversion scripts, and pushed to the hub the 3 models [here](https://huggingface.co/collections/kamilakesbi/dac-6671a8f5d1cb14dd3385d798) (16/24 and 44 khz). 
3. Made sure the forward pass gives the same output as the original model
4. Added a Feature Extractor (very similar to the Encodec FeatureExtractor). 
5. Started iterating on tests. 

# Who can review ? 

cc @sanchit-gandhi and @ArthurZucker 
cc @ylacombe for visibility

kamilakesbi committed:
dac model

kamilakesbi committed:
original dac works

kamilakesbi committed:
add dac model

kamilakesbi committed:
dac can be instatiated

kamilakesbi committed:
add forward pass

kamilakesbi committed:
load weights

kamilakesbi committed:
all weights are used

kamilakesbi committed:
convert checkpoint script ready

kamilakesbi committed:
test

kamilakesbi committed:
add feature extractor

kamilakesbi committed:
up

kamilakesbi committed:
make style

kamilakesbi committed:
apply cookicutter

kamilakesbi committed:
fix tests

kamilakesbi committed:
iterate on FeatureExtractor

kamilakesbi committed:
nit

kamilakesbi committed:
update dac doc

kamilakesbi committed:
replace nn.Sequential with nn.ModuleList

kamilakesbi committed:
nit

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,616 @@
+    """
+    ResidualVectorQuantize block - Introduced in SoundStream: An end2end neural audio codec (https://arxiv.org/abs/2107.03312)
+    """
+    def __init__(
```
Fits on one-line:
```suggestion
    def __init__(self, config: DacConfig):
```

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,616 @@
+    ):
+        super().__init__()
+
+        input_dim = config.latent_dim
```
There's no point here re-assigning all these config attributes to new variable names -> the config should pass directly to  `DacVectorQuantize`, saving you all this boilerplate

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,616 @@
+        codebook = F.normalize(codebook)
+
+        # Compute euclidean distance with codebook
+        dist = (
```
This is a bit cluttered. Could we clean it with something like:
```suggestion
        l2_norm = encodings.pow(2).sum(1, keepdim=True)
        dist = (l2_norm - 2 * encodings @ codebook.t()) + l2_norm.t()
```

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,616 @@
+            projected_latents_i = self.quantizers[i].decode_code(codebook_indices[:, i, :])
+            projected_latents.append(projected_latents_i)
+
+            quantized_representation_i = self.quantizers[i].out_proj(projected_latents_i)
```
Can be simplified:
```suggestion
            quantized_representation += self.quantizers[i].out_proj(projected_latents_i)
```

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,616 @@
+        projected_latents = []
+
+        if n_quantizers is None:
+            n_quantizers = self.n_codebooks
```
(nit) usually we do single line `if`'s like this on one-line:
```suggestion
        n_quantizers = n_quantizers if n_quantizers is not None else self.n_codebooks
```

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,616 @@
+            if self.training is False and i >= n_quantizers:
+                break
+
+            quantized_representation_i, commitment_loss_i, codebook_loss_i, indices_i, z_e_i = quantizer(residual)
```
`z_e_i` is not a great name, otherwise this module looks good!

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,616 @@
+    def forward(self, input):
+        for model in self.model: 
+            input = model(input)
+        return input
```
Should return an instance of `DacDecoderOutput`

kamilakesbi replied:
Should we return an instance of `DacDecoderOutput` in `DacModel.decode` ?

sanchit-gandhi replied:
Yep that's right!

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,616 @@
+
+        length = input_values.shape[-1]
+        encoder_output = self.encode(input_values, n_quantizers)
+        x = self.decode(encoder_output.z)
```
`z` and `x` not great names

sanchit-gandhi commented on src/transformers/models/dac/configuration_dac.py:
```diff
@@ -0,0 +1,49 @@
+
+    def __init__(
+        self,
+        encoder_dim=64,
```
`encoder_hidden_size` would be more consistent

sanchit-gandhi commented on src/transformers/models/dac/configuration_dac.py:
```diff
@@ -0,0 +1,49 @@
+    def __init__(
+        self,
+        encoder_dim=64,
+        encoder_rates=[2, 4, 8, 8],
```
Could we just have one argument:
```python
    upsampling_ratios=[2, 4, 8, 8],
```

And then set `downsampling_ratios = upsampling_ratios[::-1]` in the modelling code?

sanchit-gandhi commented on src/transformers/models/dac/configuration_dac.py:
```diff
@@ -0,0 +1,49 @@
+        self,
+        encoder_dim=64,
+        encoder_rates=[2, 4, 8, 8],
+        latent_dim=None,
```
How is this different from `encoder_hidden_size`? If required, can we re-name it to `hidden_size` and set a suitable default?

kamilakesbi replied:
It is computed this way: `latent_dim = encoder_hidden_size * (2 ** len(downsampling_ratios))`. 

We can remove it from the config parameters, rename it `hidden_size`, and infer it using `encoder_hidden_size` and `downsampling_ratios`.

sanchit-gandhi commented on src/transformers/models/dac/configuration_dac.py:
```diff
@@ -0,0 +1,49 @@
+
+logger = logging.get_logger(__name__)
+
+class DacConfig(PretrainedConfig):
```
Missing docstrings!

sanchit-gandhi commented on src/transformers/models/dac/configuration_dac.py:
```diff
@@ -0,0 +1,49 @@
+        codebook_size=1024,
+        codebook_dim=8,
+        quantizer_dropout=False,
+        sample_rate=44100,
```
Not used
```suggestion
```

sanchit-gandhi commented on src/transformers/models/dac/configuration_dac.py:
```diff
@@ -0,0 +1,49 @@
+        n_codebooks=9,
+        codebook_size=1024,
+        codebook_dim=8,
+        quantizer_dropout=False,
```
Should this be a bool? Shouldn't it be a dropout factor?

kamilakesbi replied:
In the original repository it is defined as a `bool` (see [here](https://github.com/descriptinc/descript-audio-codec/blob/c7cfc5d2647e26471dc394f95846a0830e7bec34/dac/model/dac.py#L158)) and set to `False` in the checkpoints metadata. 

It does makes sense to replace it with a float between 0 and 1, and set the default to 0 tough.

sanchit-gandhi commented on src/transformers/models/dac/feature_extraction_dac.py:
```diff
@@ -0,0 +1,184 @@
+        input_values = BatchFeature({"input_values": raw_audio})
+
+        # normal padding on batch
+        padded_inputs = self.pad(
```
Why do we pad here and then pad again 3 lines later?

kamilakesbi replied:
It's probably not the best way of doing this here. 

1. The first padding is used to pad the batch of audio inputs and copy the logic from the `EncodecFeatureExtractor`. 
2. The second padding is used to expand the batch of audio inputs to a multiple of the `hop length', as is done [in the DAC preprocessing method](https://github.com/descriptinc/descript-audio-codec/blob/c7cfc5d2647e26471dc394f95846a0830e7bec34/dac/model/dac.py#L198).

sanchit-gandhi replied:
Right I see! Could we instead do one `.pad` operation, but pass `pad_to_multiple_of=self.hop_length`? _c.f._ the corresponding [docs](https://huggingface.co/docs/transformers/en/main_classes/&amp;num;transformers.SequenceFeatureExtractor.pad.pad_to_multiple_of)

sanchit-gandhi commented on src/transformers/models/dac/feature_extraction_dac.py:
```diff
@@ -0,0 +1,184 @@
+
+    # This is a property because you might want to change the chunk_length_s on the fly
+    @property
+    def chunk_length(self) -> Optional[int]:
```
Can remove these un-used properties

sanchit-gandhi commented on src/transformers/models/dac/feature_extraction_dac.py:
```diff
@@ -0,0 +1,184 @@
+            If defined the audio is pre-processed into chunks of lengths `chunk_length_s` and then encoded.
+        hop_length (`int`, *optional*, defaults to 512):
+            Overlap length between successive windows. 
+        overlap (`float`, *optional*):
```
Can then remove un-used args

sanchit-gandhi commented on src/transformers/models/encodec/convert_encodec_checkpoint_to_pytorch.py:
```diff
@@ -240,6 +239,7 @@ def recursively_load_weights(orig_dict, hf_model, model_name):
                 elif "weight_v" in name:
                     weight_type = "weight_v"
                 elif "weight_ih_l0" in name:
+                    print("ok")
```
Remember to remove changes in this file before merge

sanchit-gandhi reviewed (commented):
Mainly reviewed the modelling and configuration code, with a quick glance over the feature extraction code. Leaving a checklist here that summarises the review:

### Modelling

#### Variable naming
- [ ] Input and output names should be consistent with other models in the library (e.g. we should return `audio_values` instead of `audio`)
- [x] Single letter variable names should be avoided (e.g. `hidden_states` instead of `x`)
- [x] Latent variables should unanimously be called `hidden_states` for consistency with other models

#### Module Instantiation
- [ ] All modules should take the config as input to the `__init__` method and use this to determine all attributes where possible. You can then limit the number of ad-hoc args to modules
- [x] Reserve `nn.ModuleList` for when we have a sequence of repeated modules (e.g. 4 decoder layers). Otherwise, explicitly write out the module classes


#### Module forward
- [x] We can remove one-line methods (e.g. `embed_code`) in favour of writing them out explicitly where used
- [x] Remove all `einops.rearrange` in favour of `torch.reshape`

#### Module output
- [ ] We're currently missing the `return_dict` arg: can you copy the logic from EnCodec?
- [ ] Output of the decoder model should be an instance of `DacDecoderOutput`, not a torch tensor
- [ ] We currently return two losses: `commitment_loss` and `codebook_loss`. For compatibility with the trainer, we need to return `loss` at the first position in the output class (see [Seq2SeqLMOutput](https://huggingface.co/docs/transformers/en/main_classes/output#transformers.modeling_outputs.Seq2SeqLMOutput)). Do they weight these two losses in the original code? If so, we can add a config attribute (`commitment_weight` or something similarly appropriate) that sets a default weight, and always return the loss in the first position
- [ ] Typically we need use reference labels when computing the loss, e.g. as [for Whisper](https://github.com/huggingface/transformers/blob/1e79eade41c346d1d8c2ad79c9e32e3186f20b54/src/transformers/models/whisper/modeling_whisper.py#L1773). Here, we always seem to return the loss in the encoder -> could we unify the logic with other models?

### Configuration
- [ ] Same as the input/outputs to the model, let's try and unify the argument names with other models in the lib

kamilakesbi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,616 @@
+            improves training stability
+    """
+
+    def __init__(self, input_dim: int, codebook_size: int, codebook_dim: int):
```
Yes, the only thing is that we need to specify which `codebook_dim` to use: https://github.com/huggingface/transformers/blob/47c24ec426fa0b2389c0dbdd9453ca5c4be236e5/src/transformers/models/dac/modeling_dac.py#L304

kamilakesbi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,616 @@
+        super().__init__()
+        self.alpha = nn.Parameter(torch.ones(1, channels, 1))
+
+    def forward(self, input):
```
Torchscript expects all variables to be defined inside or as input to the method with `@torch.jit.script` decorator. Here we can't use `self.alpha' inside the forward method. 

Two solutions: 

1. We define the class without the `@torch.jit.script` decorator on the forward method.
2. We use `torch.jit.script` to script the entire module. 

--> I choose option 2 for now.

sanchit-gandhi replied:
Okay sounds good for now. Does it really bring 1.4x speed-up? Would be in favour of keeping it with option 2 if so!

kamilakesbi replied:
I think so yes :)

kamilakesbi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,616 @@
+
+        return quantized_representation, commitment_loss, codebook_loss, codebook_indices, projected_latents
+
+    def embed_code(self, embed_id):
```
You're right, and I don't think it's the case from the [official DAC training script](https://github.com/descriptinc/descript-audio-codec/blob/main/scripts/train.py).

kamilakesbi committed:
Update src/transformers/models/dac/modeling_dac.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

kamilakesbi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,616 @@
+
+
+@dataclass
+class DacDecoderOutput(ModelOutput):
```
Yes we should! I've updated the code.

kamilakesbi committed:
apply review suggestions 1/2

kamilakesbi committed:
up

kamilakesbi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,616 @@
+
+class DacEncoderBlock(nn.Module):
+    """Encoder block used in DAC encoder."""
+    def __init__(self, dimension: int = 16, stride: int = 1):
```
Not necessarily: we double the `dimension` for each `DacEncoderBlock` in [this for loop](https://github.com/huggingface/transformers/blob/70f7b3fc2da96e0b8dd0107642da05b1dbc1eff5/src/transformers/models/dac/modeling_dac.py#L468), but the dimension size can't be directly computed from the stride. (Example: for `dac_16khz`, `strides` are equal to `[2, 4, 5, 8]`.). 

--> One possibility would be to pass `config`, `stride` and `stride_position` (the index position of `stride` in `strides`) as input to `DacEncoderBlock`.

sanchit-gandhi replied:
SGTM!

kamilakesbi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,616 @@
+        quantized_representation = 0
+        quantized_latents = []
+        codes = []
+        dims = np.cumsum([0] + [q.codebook_dim for q in self.quantizers])
```
Actually the method `from_latents` isn't used by the forward pass! does it make sense to keep it ?

kamilakesbi committed:
apply review suggestions 2/2

kamilakesbi committed:
update padding in FeatureExtractor

kamilakesbi commented:
They indeed add weights to the different losses during training ([see original codebase](https://github.com/descriptinc/descript-audio-codec/blob/c7cfc5d2647e26471dc394f95846a0830e7bec34/scripts/train.py#L364)). I'll add weight attributes in the config file.  

Note that in the current code, we only return the `commitment_loss` and `codebook_loss`, but there are other losses used to train the model (`mel_loss` and gan losses).

kamilakesbi committed:
apply review suggestions

kamilakesbi committed:
iterate on design and tests

kamilakesbi committed:
add integration tests

kamilakesbi committed:
feature extractor tests

kamilakesbi committed:
make style

kamilakesbi committed:
all tests pass

kamilakesbi committed:
make style

kamilakesbi committed:
fixup

kamilakesbi commented:
I took all of Sanchit's reviews and added integration tests. @ylacombe this should be ready for review! 

- Note that there's still one failing test which indicates: 

> 1 failed because `AssertionError` -> <class 'transformers.models.dac.modeling_dac.DacModel'> is too big for the common tests (74175906)! It should have 1M max.

Should we overwrite this common test in the dac test file ? 

- There's also apparently a problem with the dosctring in DacConfig: 

> ValueError: There was at least one problem when checking docstrings of public objects.
> The argument part of the docstrings of the following objects could not be processed, check they are properly formatted.
> - DacConfig

I'm not sure what the problem is here either. 

cc @sanchit-gandhi

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31494). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

sanchit-gandhi commented on tests/models/dac/test_modeling_dac.py:
```diff
@@ -0,0 +1,563 @@
+        batch_size=5,
+        num_channels=1,
+        is_training=False,
+        intermediate_size=15992,
```
The hidden-dims used here are way too big for the fast tests. We should use the smallest possible version of the model that lets us run these tests. This ensures the fast tests actually run **fast**! Remember that there are over 200 models in the Transformers library -> we need each model tester to run in << 30-seconds to keep the speed of the CI reasonable. This is primarily achieved by using super small models and the lowest possible batch sizes.

See the dims used in EnCodec as an example of the order of magnitude these dims should be: https://github.com/huggingface/transformers/blob/5e89b335ab088b52e5dbb50c5368d944e9052ed4/tests/models/encodec/test_modeling_encodec.py#L76

kamilakesbi replied:
Oh ok I see, that makes sense :)

ylacombe commented on src/transformers/models/dac/feature_extraction_dac.py:
:fire:

ylacombe commented on tests/models/dac/test_feature_extraction_dac.py:
```diff
@@ -0,0 +1,204 @@
+
+        return [x["array"] for x in audio_samples]
+
+    def test_integration(self):
```
Could we also add a test that feature extraction also matches at the end of the audio since we're padding to multiple of smthg ?
cf this [discussion](https://github.com/huggingface/transformers/pull/31494/files#r1648768547)

ylacombe commented on tests/models/dac/test_feature_extraction_dac.py:
```diff
@@ -0,0 +1,204 @@
+# coding=utf-8
+# Copyright 2021-2023 HuggingFace Inc.
```
```suggestion
# Copyright 2024 HuggingFace Inc.
```

ylacombe commented on tests/models/dac/test_modeling_dac.py:
```diff
@@ -0,0 +1,563 @@
+# coding=utf-8
+# Copyright 2023 The HuggingFace Inc. team. All rights reserved.
```
(nit)

```suggestion
# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
```

ylacombe commented on tests/models/dac/test_modeling_dac.py:
```diff
@@ -0,0 +1,563 @@
+
+
+@require_torch
+class DacModelTester:
```
It's seems to be more or less the same tests as Encodec! 
Could you add a [`# Copied ...`](https://huggingface.co/docs/transformers/v4.42.0/en/add_new_model#code-style) from statement and some `# Ignore copy` at the right place, to facilitate the core maintainer review ?

ylacombe commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,653 @@
+    projected_latents: torch.FloatTensor = None
+
+
+@dataclass
+class DacDecoderOutput(ModelOutput):
+    """
+    Args:
+        audio_values (`torch.Tensor` of shape `(batch_size, 1, input_length)`):
+            Decoded audio data.
+    """
+
+    audio_values: torch.FloatTensor = None
```
I see that it's similar to `EncodecDecoderOutput` but with very few differences. In general, note that in transformers, we seek to minimize new pieces of code whenever it's possible.

This lead me to two comments / questions:
1. why do we return a 3D tensor instead of a 2D tensor of shape `(batch_size, input_length)` ?
2. If you finally decide to go for a 2D tensor, you could simply use a Copied from statement like this one: 

For example, here, you could add:

```suggestion
@dataclass
# Copied from transformers.models.encodec.modeling_encodec.EncodecDecoderOutput with Encodec->Dac, segment_length->input_length
class DacDecoderOutput(ModelOutput):
    """
    Args:
        audio_values (`torch.FloatTensor`  of shape `(batch_size, input_length)`, *optional*):
            Decoded audio values, obtained using the decoder part of Dac.
    """

    audio_values: torch.FloatTensor = None
```

kamilakesbi replied:
We can definitely return a 2D tensor! we returned a 3D tensor mostly to mimic the original model behaviour.

ylacombe commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,653 @@
+    def __init__(self, config: DacConfig):
+        super().__init__()
+
+        self.in_proj = weight_norm(nn.Conv1d(config.hidden_size, config.codebook_dim, kernel_size=1))
+        self.out_proj = weight_norm(nn.Conv1d(config.codebook_dim, config.hidden_size, kernel_size=1))
```
If I'm not mistaken, we usually apply `weight_norm` once when converting !
See how it's done for SpeechT5:
1. create utility methods to apply / remove weight norm (cf [here](https://github.com/huggingface/transformers/blob/e65502951593a76844e872fee9c56b805598538a/src/transformers/models/speecht5/modeling_speecht5.py#L3312-L3326))
2. when converting the weights, [apply the weight norm](https://github.com/huggingface/transformers/blob/e65502951593a76844e872fee9c56b805598538a/src/transformers/models/speecht5/convert_hifigan.py#L30) before loading the state dict then [remove the weight norm ](https://github.com/huggingface/transformers/blob/e65502951593a76844e872fee9c56b805598538a/src/transformers/models/speecht5/convert_hifigan.py#L55)when conversion is done


cf @amyeroberts, I just want to double check that this is correct!

kamilakesbi replied:
mmh I see, that's how it's defined in the original codebase, so I wonder if we shouldn't just leave it as it is... would also like to hear @amyeroberts opinion!

amyeroberts replied:
I` would stick with how it's been done in other models in transformers, i.e. as suggested by @ylacombe. Even if it's done a certain way in the original repo, it's better to have consistency and part of bringing a model into the library it making these conversions

ylacombe commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,653 @@
+        pad = ((7 - 1) * dilation) // 2
+
+        self.snake1 = Snake1d(dimension)
+        self.conv1 = weight_norm(nn.Conv1d(dimension, dimension, kernel_size=7, dilation=dilation, padding=pad))
```
Same comment for the weight_norm, here and in the rest of the modeling code !

ylacombe commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,653 @@
+
+        return DacEncoderOutput(loss, quantized_representation, codebook_indices, projected_latents)
+
+    def decode(
+        self,
+        quantized_representation: torch.Tensor,
+        return_dict: Optional[bool] = None,
+    ):
+        """Decode given latent codes and return audio data
+
+        Args:
+            quantized_representation (torch.Tensor of shape `(batch_size, dimension, time_steps)`):
+                Quantized continuous representation of input
+            return_dict (`bool`, *optional*):
+                Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+        Returns:
+            audio_values (`torch.Tensor` of shape `(batch_size, 1, input_length)`):
+                Decoded audio data.
+        """
+        return_dict = return_dict if return_dict is not None else self.config.return_dict
+
+        audio_values = self.decoder(quantized_representation)
+
+        if not return_dict:
+            return (audio_values,)
+
+        return DacDecoderOutput(audio_values)
```
We should absolutely also be able to also decode from `audio_codes`, as it'll be the most common use-case !

I think it's done here in the original codebase: https://github.com/descriptinc/descript-audio-codec/blob/c7cfc5d2647e26471dc394f95846a0830e7bec34/dac/model/base.py#L236-L294

ylacombe replied:
This is also what is done in Encodec FYI ! https://github.com/sanchit-gandhi/parler-tts/blob/10016fb0300c0dc31a0fb70e26f3affee7b62f16/parler_tts/dac_wrapper/modeling_dac.py#L98-L134

ylacombe commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,653 @@
+    def __init__(self, config: DacConfig):
+        super().__init__()
+
+        # d_model = config.encoder_hidden_size
```
(nit)
to remove
```suggestion
```

ylacombe commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,653 @@
+    def forward(self, hidden_state):
+        hidden_state = self.conv1(hidden_state)
+
+        for module in self.block:
+            hidden_state = module(hidden_state)
```
Any particular reason we iterate over self.block instead of simply calling `hidden_state = self.block(hidden_state)` ?

kamilakesbi replied:
because `self.block` is a [ModuleList](https://pytorch.org/docs/stable/generated/torch.nn.ModuleList.html)!

ylacombe commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,653 @@
+    def forward(self, hidden_state):
+        hidden_state = self.conv1(hidden_state)
+
+        for layer in self.block:
+            hidden_state = layer(hidden_state)
```
Same comment here !

ylacombe commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,653 @@
+    loss: torch.FloatTensor = None
+    audio_values: torch.FloatTensor = None
+    quantized_representation: torch.FloatTensor = None
+    codebook_indices: torch.FloatTensor = None
```
Let's try to keep the same naming as Encodec, at least for the codebook indices ! 

Let's do [`audio_codes`](https://github.com/huggingface/transformers/blob/e65502951593a76844e872fee9c56b805598538a/src/transformers/models/encodec/modeling_encodec.py#L67) instead of `codebook_indices` ? WDYT ?

kamilakesbi replied:
LGTM!

ylacombe commented on docs/source/en/model_doc/dac.md:
I think you can start writing this doc since the PR is already in great shape! 
You can take a look at Encodec doc for inspiration ! https://github.com/huggingface/transformers/blob/3345ae733b6f4aeb7204a0f3e646a3cdbaad0023/docs/source/en/model_doc/encodec.md#L4

ylacombe reviewed (commented):
Hey @kamilakesbi,
Thanks for this great PR !!

I've left a few comments, the main ones being:
1. we definitely should have a method to decode from audio codebooks. We could maybe make the `decode` method compatible with both audio codebooks and quantized representation, WDYT ?
2. I'm not quite sure that we should have the losses being computed by default, especially since these losses are alone not enough to train the model - we need a few more losses to train the model if I remember correctly !

Let me know if you got any further questions, but again congrats on the PR, it's looking really great!

kamilakesbi committed:
apply review suggestions

kamilakesbi committed:
fix-copies

kamilakesbi commented:
Thank you for your review @ylacombe! 

I have taken your feedback into account and updated the code. I've also added the ability to decode from audio codebooks. 

Regarding the loss, I agree with you that we should probably not return the encoder loss by default. Normally, the `loss` is returned when the `labels` argument is passed to the forward pass of the model, but here I'm not sure it makes sense to add a `labels` argument, as the loss is computed in an unsupervised manner. Should I add a `return_loss` arg instead ? 

Otherwise I think this is ready for a final review @amyeroberts :) failling tests are unrelated to this PR I think.

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,665 @@
+        loss (`torch.Tensor`):
+            Loss from the encoder model, comprising the weighted combination of the commitment and codebook losses.
+        audio_values (`torch.Tensor` of shape `(batch_size, input_length)`):
+            Decoded audio data.
```
```suggestion
            Reconstructed audio data.
```

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,665 @@
+    loss: torch.FloatTensor = None
+    audio_values: torch.FloatTensor = None
+    quantized_representation: torch.FloatTensor = None
+    audio_codes: torch.FloatTensor = None
```
Are the audio codes a float tensor or long tensor of ints?

kamilakesbi replied:
It's a long tensor of ints indeed

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,665 @@
+    audio_values: torch.FloatTensor = None
+
+
+class Snake1d(nn.Module):
```
In the end it looks like `torch.jit` was removed entirely. Any reason for not going with option 2 discussed in the previous review? https://github.com/huggingface/transformers/pull/31494#discussion_r1649135930

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,665 @@
+            hidden_state (`torch.Tensor` of shape `(batch_size, dimension, time_steps)`):
+                Input tensor.
+
+        Returns:
```
This docstring looks to be a bit weirdly formatted (missing some spaces, inconsistent formatting) - could you check you follow the same format for all entries? Specifically:
```python
variable_name: (`torch.{Long, Float}Tensor of shape `(...)`):
    Variable description.
```

amyeroberts replied:
This wasn't resolved

amyeroberts replied:
This still wasn't resolved. Please make sure to fully address all reviewers' comments or comment on why the suggestion isn't being applied.

kamilakesbi replied:
sorry I didn't see it... should be good now :)

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,665 @@
+        projected_latents = self.in_proj(hidden_state)
+        quantized_representation, audio_codes = self.decode_latents(projected_latents)
+
+        commitment_loss = F.mse_loss(projected_latents, quantized_representation.detach(), reduction="none").mean(
```
Would we get the same result by doing:
```python
commitment_loss = F.mse_loss(projected_latents, quantized_representation.detach(), reduction="mean")
```

kamilakesbi replied:
yes indeed :)

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,665 @@
+        l2_norm = encodings.pow(2).sum(1, keepdim=True)
+        dist = (l2_norm - 2 * encodings @ codebook.t()) + codebook.pow(2).sum(1, keepdim=True).t()
+
+        indices = (-dist).max(1)[1].reshape(hidden_states.size(0), -1)
```
(nit) maybe put the `-` sign on the line before so that we can do:
```python
indices = dist.max(1)[1]
indices = indices.reshape(hidden_states.size(0), -1)
```
Which is marginally cleaner than what's currently written

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,665 @@
+                Number of quantizers to use. If specified and `self.quantizer_dropout` is True,
+                this argument is ignored during training, and a random number of quantizers is used.
+
+        Returns:
```
In the `Returns` portion of the docstring, nothing is "*optional*", so we can remove these!

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,665 @@
+        )
+        audio_values = self.decode(quantized_representation, return_dict=False)
+
+        if isinstance(audio_values, tuple):
```
Why would `audio_values` be a tuple here?

kamilakesbi replied:
The output to `decode` will be a tuple if return_dict is set to False. Actually in the `forward` method `audio_values` is always a tuple! (no need to check if it's a tuple or not... I'll update the code accordingly).

sanchit-gandhi commented on docs/source/en/model_doc/dac.md:
```diff
@@ -0,0 +1,69 @@
+
+-->
+
+# dac
```
Can title-case for the doc
```suggestion
DAC
```

sanchit-gandhi commented on docs/source/en/model_doc/dac.md:
```diff
@@ -0,0 +1,69 @@
+
+## Overview
+
+The dac model was proposed in [Descript Audio Codec: High-Fidelity Audio Compression with Improved RVQGAN](https://arxiv.org/abs/2306.06546) by <Rithesh Kumar, Prem Seetharaman, Alejandro Luebs, Ishaan Kumar, Kundan Kumar>.
```
```suggestion
The DAC model was proposed in [Descript Audio Codec: High-Fidelity Audio Compression with Improved RVQGAN](https://arxiv.org/abs/2306.06546) by Rithesh Kumar, Prem Seetharaman, Alejandro Luebs, Ishaan Kumar, Kundan Kumar.
```

sanchit-gandhi commented on docs/source/en/model_doc/dac.md:
```diff
@@ -0,0 +1,69 @@
+
+*Language models have been successfully used to model natural signals, such as images, speech, and music. A key component of these models is a high quality neural compression model that can compress high-dimensional natural signals into lower dimensional discrete tokens. To that end, we introduce a high-fidelity universal neural audio compression algorithm that achieves ~90x compression of 44.1 K ...
+
+Tips:
```
```suggestion

```

sanchit-gandhi commented on docs/source/en/model_doc/dac.md:
```diff
@@ -0,0 +1,69 @@
+Tips:
+
+This model was contributed by [Kamil Akesbi](https://huggingface.co/kamilakesbi).
+The original code can be found [here](https://github.com/descriptinc/descript-audio-codec/tree/main?tab=readme-ov-file.
```
```suggestion
The original code can be found [here](https://github.com/descriptinc/descript-audio-codec/tree/main?tab=readme-ov-file).
```

sanchit-gandhi commented on docs/source/en/model_doc/dac.md:
```diff
@@ -0,0 +1,69 @@
+
+## Usage example 
+
+Here is a quick example of how to encode and decode an audio using this model:
```
Encoding + decoding is a nice way of showcasing model performance, but the typical practitioner use-case is encoding to get the audio codebook representation - could we adjust the example to highlight how a user can do this, e.g. as an intermediate step in the example?

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,665 @@
+
+        padding = (hidden_state.shape[-1] - output_tensor.shape[-1]) // 2
+        if padding > 0:
+            hidden_state = hidden_state[..., padding:-padding]
```
We're sure this is correct? No need for a padding mask, just equal left/right padding on the downsampled hidden states?

kamilakesbi replied:
This is how it's done in the original codebase...

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,665 @@
+        n_quantizers: int = None,
+        return_dict: Optional[bool] = None,
+    ):
+        """
```
Would be nice to add an example to this `forward` docstring, e.g. in the same way it's done in EnCodec

sanchit-gandhi commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,665 @@
+# coding=utf-8
```
Overall this modelling file looks much cleaner!

sanchit-gandhi commented on src/transformers/models/dac/feature_extraction_dac.py:
```diff
@@ -0,0 +1,172 @@
+            pad_to_multiple_of=self.hop_length,
+        )
+
+        if padding is not False:
```
We can just do:
```suggestion
        if padding:
```

sanchit-gandhi commented on src/transformers/models/dac/feature_extraction_dac.py:
```diff
@@ -0,0 +1,172 @@
+        )
+
+        if padding is not False:
+            padded_inputs.input_values = padded_inputs.input_values[:, np.newaxis, :]
```
Not fully sure why we do this!

kamilakesbi replied:
The model takes as input tensors of the form `(batch_size, 1, num_samples)`, but the input_values returned by the padding method are of the form `(bs, num_samples)`.

sanchit-gandhi commented on src/transformers/models/dac/feature_extraction_dac.py:
```diff
@@ -0,0 +1,172 @@
+            raw_audio = [np.asarray(raw_audio).T]
+
+        # verify inputs are valid
+        for idx, example in enumerate(raw_audio):
```
DAC only has mono checkpoints right? In this case, we only need to check if the input is mono. No need to include the stereo logic

sanchit-gandhi commented on tests/models/dac/test_feature_extraction_dac.py:
```diff
@@ -0,0 +1,206 @@
+
+
+@require_torch
+class DacFeatureExtractionTest(SequenceFeatureExtractionTestMixin, unittest.TestCase):
```
This test suite looks quite similar to the EnCodec feature extraction one - any chance of `# Copied from` statements here?

sanchit-gandhi commented on tests/models/dac/test_modeling_dac.py:
```diff
@@ -0,0 +1,568 @@
+@slow
+@require_torch
+class DacIntegrationTest(unittest.TestCase):
+    def test_integration(self):
```
Very thorough! However, I would split this up into three slow tests (one test each for 16, 24 and 44 kHz). It's better that a test checks one very focussed thing (in this case a single checkpoint), rather than a huge expansive set of checks (which are much harder to debug if they start failing)

sanchit-gandhi commented on tests/models/dac/test_modeling_dac.py:
```diff
@@ -0,0 +1,568 @@
+        models = ["dac_16khz", "dac_24khz", "dac_44khz"]
+
+        for model_name in models:
+            model_id = "kamilakesbi/{}".format(model_name)
```
Remember to update to the Hub checkpoitns

kamilakesbi replied:
I'll update them as soon as we move them to the `descript` organisation!

sanchit-gandhi commented on tests/models/dac/test_modeling_dac.py:
```diff
@@ -0,0 +1,568 @@
+                rmse = compute_rmse(arr_cut, arr_enc_dec_cut)
+                self.assertTrue(rmse < expected_rmse[str(model_name)])
+
+    def test_integration_batch(self):
```
Likewise here - better to have three tests

sanchit-gandhi reviewed (approved):
Looks in great shape - just some minor style nits from me!

kamilakesbi committed:
apply review suggestions

ylacombe commented on docs/source/en/model_doc/dac.md:
```diff
@@ -0,0 +1,68 @@
+
+The abstract from the paper is the following:
+
+*Language models have been successfully used to model natural signals, such as images, speech, and music. A key component of these models is a high quality neural compression model that can compress high-dimensional natural signals into lower dimensional discrete tokens. To that end, we introduce a high-fidelity universal neural audio compression algorithm that achieves ~90x compression of 44.1 K ...
```
(nit)
```suggestion
*Language models have been successfully used to model natural signals, such as images, speech, and music. A key component of these models is a high quality neural compression model that can compress high-dimensional natural signals into lower dimensional discrete tokens. To that end, we introduce a high-fidelity universal neural audio compression algorithm that achieves ~90x compression of 44.1 KHz audio into tokens at just 8kbps bandwidth. We achieve this by combining advances in high-fidelity audio generation with better vector quantization techniques from the image domain, along with improved adversarial and reconstruction losses. We compress all domains (speech, environment, music, etc.) with a single universal model, making it widely applicable to generative modeling of all audio. We compare with competing audio compression algorithms, and find our method outperforms them significantly. We provide thorough ablations for every design choice, as well as open-source code and trained model weights. We hope our work can lay the foundation for the next generation of high-fidelity audio modeling.*
```

ylacombe commented on docs/source/en/model_doc/dac.md:
```diff
@@ -0,0 +1,68 @@
+
+## Overview
+
+The DAC model was proposed in [Descript Audio Codec: High-Fidelity Audio Compression with Improved RVQGAN](https://arxiv.org/abs/2306.06546) by <Rithesh Kumar, Prem Seetharaman, Alejandro Luebs, Ishaan Kumar, Kundan Kumar>.
```
(nit)

```suggestion
The DAC model was proposed in [Descript Audio Codec: High-Fidelity Audio Compression with Improved RVQGAN](https://arxiv.org/abs/2306.06546) by Rithesh Kumar, Prem Seetharaman, Alejandro Luebs, Ishaan Kumar, Kundan Kumar.
```

ylacombe reviewed (approved):
Thanks for iterating @kamilakesbi, LGTM!

gentle ping to  @amyeroberts and @ArthurZucker for a review!

kamilakesbi committed:
Update docs/source/en/model_doc/dac.md

Co-authored-by: Yoach Lacombe <52246514+ylacombe@users.noreply.github.com>

kamilakesbi committed:
Update docs/source/en/model_doc/dac.md

Co-authored-by: Yoach Lacombe <52246514+ylacombe@users.noreply.github.com>

amyeroberts commented on tests/models/dac/test_modeling_dac.py:
```diff
@@ -0,0 +1,750 @@
+):
+    if input_ids is not None:
+        encoder_dict = {"input_ids": input_ids}
+    else:
```
This assumes `input_values` isn't None

amyeroberts commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,683 @@
+
+        commitment_loss = F.mse_loss(projected_latents, quantized_representation.detach(), reduction="mean")
+        codebook_loss = F.mse_loss(quantized_representation, projected_latents.detach(), reduction="mean")
+        quantized_representation = (
+            projected_latents + (quantized_representation - projected_latents).detach()
+        )  # noop in forward pass, straight-through gradient estimator in backward pass
```
library convention is to put comments on the line above - this avoids unnecessary line splitting 

```suggestion
        # noop in forward pass, straight-through gradient estimator in backward pass
        quantized_representation = projected_latents + (quantized_representation - projected_latents).detach()
```

amyeroberts commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,683 @@
+    def __init__(self, config: DacConfig, stride: int = 1, stride_index: int = 1):
+        super().__init__()
+
+        input_dim = config.decoder_hidden_size // 2**stride_index
+        output_dim = config.decoder_hidden_size // 2 ** (stride_index + 1)
```
nit - consistency 

```suggestion
        input_dim = config.decoder_hidden_size // 2 ** stride_index
        output_dim = config.decoder_hidden_size // 2 ** (stride_index + 1)
```

amyeroberts commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,683 @@
+        Args:
+            hidden_state (`torch.Tensor` of shape `(batch_size, dimension, time_steps)`):
+                Input tensor to be quantized.
+            n_quantizers (`int`, optional):
```
```suggestion
            n_quantizers (`int`, *optional*):
```

amyeroberts commented on src/transformers/models/dac/modeling_dac.py:
```diff
@@ -0,0 +1,683 @@
+                to decode directly from the audio codes (it will overwrite quantized_representation).
+            return_dict (`bool`, *optional*):
+                Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+        Returns:
+            audio_values (`torch.Tensor` of shape `(batch_size, input_length)`):
+                Decoded audio data.
```
This is a bit misleading - it may return a `DacDecoderOutput` object which contains `audio_data`

amyeroberts commented on docs/source/en/model_doc/dac.md:
````diff
@@ -0,0 +1,68 @@
+
+Here is a quick example of how to encode and decode an audio using this model:
+
+```python 
````
This is a nice example. The main thing I think is missing with this model page is some context about why this is useful. I can see a good demo of being able to encode and then reconstruct but it's not clear the use case beyond that. 

The model page should have a short, clear description of the model, its uses and anything notable about the architecture in addition to the abstract. [MusicGen is a good example](https://huggingface.co/docs/transformers/v4.42.0/en/model_doc/musicgen#overview)

amyeroberts reviewed (commented):
Thanks for adding this model! 

There's a few things here and there, mainly the weight_norm logic, but overall looks really good and clean 🤗

kamilakesbi committed:
apply review suggestions

kamilakesbi committed:
anticipate transfer weights to descript

kamilakesbi commented:
Thanks for the reviews @amyeroberts and @ylacombe! 

We should be close from merging this model! 

The last change would be to transfer the weights from my [personal hugging face page](https://huggingface.co/collections/kamilakesbi/dac-6671a8f5d1cb14dd3385d798) to the [descript organisation](https://huggingface.co/descript). I'm waiting for the members of the organisation to add me.

kamilakesbi commented:
@amyeroberts the checkpoints have been transferred to the `Descript` organisation. 

We can merge this PR if everything if ok for you :)

kamilakesbi committed:
up

kamilakesbi committed:
make style

kamilakesbi commented:
Gentle ping @amyeroberts

amyeroberts commented:
@kamilakesbi There's still failing tests on the CI - these should be resolved before final review and merge. You may need to rebase on main to include upstream changes or trigger a re-run of the CI if the issues are relating to the environment or other libraries.

kamilakesbi commented:
@amyeroberts the remaining failing tests are related to the hub: 

`1 failed because huggingface_hub.utils._errors.RepositoryNotFoundError: 404 Client Error. (Request ID -> Root=1-6694ef78-6ec395936d885db72ebacd65;2c9bbc0d-6ecd-49c4-b28e-df134be7bd4a)`. 

I'm not sure exactly what's causing these tests to fail... could you help me with this? thanks!

amyeroberts reviewed (commented):
Thanks for adding and iterating! 

Just part of the feature extractor and docs to update

kamilakesbi committed:
apply review suggestions

amyeroberts reviewed (approved):
Thanks for all the work adding @kamilakesbi! 

Just two things before merge: 
* Could you rebase on `main` to include upstream changes? This should fix the hub tests
* Could you push a commit with the message `[run_slow] dac` to trigger a run of all the model's slow tests?

kamilakesbi commented:
The slow tests are passing locally but not on the CI... any idea what could be the reason here @sanchit-gandhi  @amyeroberts ?

amyeroberts commented:
@kamilakesbi From an initial glance it looks like the tests are failing on numerical checks. The model's logits are sensitive to things like the running environment and hardware, as so we expect small changes. It is weird its changing so much to be outside the tolerance but depending on what the values are it could be OK. 

The test values should be the ones for running on the CI, rather than ones own setup. You'll need to ssh into to the CI runners to get these and make sure they're not too different from the current test values

kamilakesbi committed:
update slow test values

amyeroberts commented on tests/models/dac/test_modeling_dac.py:
```diff
@@ -681,10 +681,10 @@ def test_integration_batch_44khz(self):
-            "audio_codes": 778.2920,
-            "projected_latents": -3844.6055,
+            "loss": 51.8466,
+            "quantized_representation": 595.9933,
+            "audio_codes": 2121648,
```
These are very large differences 👀  Do we expect more variation for e.g. the 44khz case?

kamilakesbi replied:
We are comparing the `sum` over values in each tensor, which I guess propagates errors and explain why we result with such a high difference. 

I switched to using `mean` instead, this seems to reduce the observed differences.

kamilakesbi committed:
update slow tests

kamilakesbi committed:
update test values

kamilakesbi committed:
update with CI values

amyeroberts commented on tests/models/dac/test_modeling_dac.py:
```diff
@@ -0,0 +1,735 @@
+        with torch.no_grad():
+            encoder_outputs = model.encode(inputs["input_values"])
+
+            expected_encoder_sums = torch.tensor(list(expected_encoder_sums_dict.values()), dtype=torch.float32)
+            encoder_outputs_sum = torch.tensor([v.float().mean().cpu().item() for v in encoder_outputs.to_tuple()])
```
Two things here: 
* The object `expected_encoder_sums_dict` should be renamed to reflect what's being tested
* If you're finding that you're unable to replicate numbers in runs, I would recommend trying just to see if you can replicate just the first set of values in the tuple for a given tolerance.

kamilakesbi committed:
update with vorace values

kamilakesbi committed:
update test with slice

kamilakesbi committed:
make style

amyeroberts commented:
@kamilakesbi Why was this merged when there were failing slow tests?
